File: typescript/examples/pro-tier/item-analytics-report/item_analytics_report.ts
import { fileURLToPath } from "node:url";
import { parseArgs } from "node:util";

import { MarketIntelligenceApi } from "cs2cap";

import { formatApiError } from "../../_shared/api.js";
import { buildConfiguration, loadApiKey } from "../../_shared/auth.js";
import { parseIntegerOption, printHelpIfRequested } from "../../_shared/cli.js";
import { formatDecimalMoney, formatPercent, renderTable } from "../../_shared/render.js";

const SCRIPT_PATH = fileURLToPath(import.meta.url);

function readOptions(): { top: number } {
  const { values } = parseArgs({
    options: {
      top: { type: "string" },
    },
  });
  return { top: parseIntegerOption(values.top, 5, "--top") };
}

function money(value: string | null | undefined): string {
  return formatDecimalMoney(value);
}

async function main(): Promise<number> {
  if (printHelpIfRequested("Usage: npm run example -- examples/pro-tier/item-analytics-report/item_analytics_report.ts [--top 5]")) {
    return 0;
  }
  const args = readOptions();

  let bearerToken: string;
  try {
    bearerToken = loadApiKey("CS2C_API_KEY", SCRIPT_PATH);
  } catch (error) {
    console.error(error instanceof Error ? error.message : String(error));
    return 1;
  }

  const configuration = buildConfiguration(bearerToken);

  try {
    const marketApi = new MarketIntelligenceApi(configuration);
    const snapshot = await marketApi.getMarketAnalyticsSnapshot();
    const candidates = [...snapshot.data.items]
      .sort((left, right) => (right.summary.liquidity ?? 0) - (left.summary.liquidity ?? 0))
      .slice(0, Math.max(args.top, 1));
    const details = await Promise.all(
      candidates.map((item) => marketApi.getItemAnalytics({ itemId: item.itemId })),
    );

    console.log("Market snapshot candidates (/v1/market/items):");
    console.log(
      renderTable(
        ["rank", "item_id", "item", "liquidity", "ask", "bid", "spread"],
        candidates.map((item) => [
          item.summary.rank ?? "N/A",
          item.itemId,
          item.marketHashName,
          item.summary.liquidity ?? "N/A",
          money(item.summary.bestAskUsd),
          money(item.summary.bestBidUsd),
          formatPercent(item.summary.avgSpreadPct),
        ]),
      ),
    );
    console.log();

    console.log("Detail coverage (/v1/market/items/{item_id}):");
    console.log(
      renderTable(
        ["item_id", "providers", "volume_providers", "bid_side_providers", "provider_rows"],
        details.map((response) => [
          response.data.itemId,
          response.data.coverage.providerCount,
          response.data.coverage.providersWithVolume,
          response.data.coverage.providersWithBidSide,
          response.data.providers.length,
        ]),
      ),
    );
    return 0;
  } catch (error) {
    console.error(await formatApiError(error));
    return 1;
  }
}

void main().then((code) => {
  process.exitCode = code;
});

File: python/examples/pro-tier/item-analytics-report/item_analytics_report.py
import argparse
import sys
from pathlib import Path

import cs2cap
from cs2cap.rest import ApiException

EXAMPLES_ROOT = Path(__file__).resolve().parents[2]
if str(EXAMPLES_ROOT) not in sys.path:
    sys.path.insert(0, str(EXAMPLES_ROOT))

from _shared.auth import build_configuration, load_api_key  # noqa: E402
from _shared.render import format_decimal_money, format_percent, render_table  # noqa: E402


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Use /v1/market/items to select candidates, then inspect detail rows.",
    )
    parser.add_argument("--top", type=int, default=5, help="Number of snapshot rows to inspect.")
    return parser


def money(value: str | None) -> str:
    return format_decimal_money(value)


def main() -> int:
    args = build_parser().parse_args()

    try:
        bearer_token = load_api_key("CS2C_API_KEY", start=Path(__file__))
    except RuntimeError as exc:
        print(str(exc), file=sys.stderr)
        return 1

    configuration = build_configuration(bearer_token)

    try:
        with cs2cap.ApiClient(configuration) as client:
            market_api = cs2cap.MarketIntelligenceApi(client)
            snapshot = market_api.get_market_analytics_snapshot()
            candidates = sorted(
                snapshot.data.items,
                key=lambda item: item.summary.liquidity or 0,
                reverse=True,
            )[: max(args.top, 1)]
            details = [
                market_api.get_item_analytics(item.item_id).data
                for item in candidates
            ]

    except ApiException as exc:
        print(f"API request failed: {exc}", file=sys.stderr)
        return 1

    print("Market snapshot candidates (/v1/market/items):")
    print(
        render_table(
            ["rank", "item_id", "item", "liquidity", "ask", "bid", "spread"],
            [
                [
                    str(item.summary.rank or "N/A"),
                    str(item.item_id),
                    item.market_hash_name,
                    str(item.summary.liquidity or "N/A"),
                    money(item.summary.best_ask_usd),
                    money(item.summary.best_bid_usd),
                    format_percent(item.summary.avg_spread_pct),
                ]
                for item in candidates
            ],
        )
    )
    print()

    print("Detail coverage (/v1/market/items/{item_id}):")
    print(
        render_table(
            ["item_id", "providers", "volume_providers", "bid_side_providers", "provider_rows"],
            [
                [
                    str(detail.item_id),
                    str(detail.coverage.provider_count),
                    str(detail.coverage.providers_with_volume),
                    str(detail.coverage.providers_with_bid_side),
                    str(len(detail.providers)),
                ]
                for detail in details
            ],
        )
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
